### PR TITLE
refactor and bump to v1.1.0

### PR DESCRIPTION
**Pull request by zhmushan**
- [x] rm node_modules
- [x] using semvers version range syntax.
- [x] add tests
- [x] ci [win, ubuntu, mac]

### CODE REVIEW

**Comment by zhmushan** [comment]
Any ideas to get all available versions of deno without github api? it caused macos ci red.
@axetroy

**Comment by axetroy** [comment]
> Any ideas to get all available versions of deno without github api? it caused macos ci red.

here is the [API for github](https://developer.github.com/v3/repos/releases/#get-the-latest-release)

https://api.github.com/repos/denoland/deno/releases/latest

**Comment by zhmushan** [comment]
@axetroy https://github.com/zhmushan/setup-deno/runs/321272230
image

**Comment by axetroy** [comment]
@zhmushan 

This is the interface limitation of the Github API.

We can add personal token in `Github Secret` and use it in pipeline.

But I don’t approve of it.

Now I am referring to how setup-deno is implemented

**Comment by axetroy** [comment]
Here is the `setup-node` does

https://github.com/actions/setup-node/blob/9a99bb3e353525a0754b4b99d40dfd7fc812cae9/src/installer.ts#L101-L106

It provides a version map file.

This is what we should do

**Comment by axetroy** [comment]
We can temporarily store this file in the repository directory

After merging the code, move it to `gh-pages`

This will be accessible via `https://setup-deno.denolib.io/release.json`

**Comment by axetroy** [comment]
@zhmushan 

maybe put that file(release.json) into `/docs` will be better

No need to switch to `gh-pages` branch.

**Comment by zhmushan** [comment]
@axetroy ready for review

File: src/installer.ts [hunk 1]
```diff
@@ -1,137 +1,166 @@
...
-    "deno" + executableExtension,
-    "deno",
-    version
-  );
-}
+import * as os from "os";
+import * as path from "path";
+
+// On load grab temp directory and cache directory and remove them from env (currently don't want to expose this)
+let tempDirectory: string = process.env["RUNNER_TEMP"] || "";
```
Comment (axetroy) [inline src/installer.ts]: Is there any reason to put this before `import`?

File: src/release.json [hunk 2]
```diff
@@ -0,0 +1,302 @@
+[
+  {
+    "name": "v0.25.0",
```
Comment (axetroy) [inline src/release.json]: Can you describe how this file was produced

It's better to have a documentation

File: src/installer.ts (hunk 1 above)
Comment (zhmushan) [reply src/installer.ts]: https://github.com/actions/toolkit/blob/master/packages/tool-cache/src/tool-cache.ts#L24
tool-cache gets tempDir from env, so we need to set env before import

File: __tests__/installer.test.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  path.join(
+    Math.random()
+      .toString(36)
+      .substring(7)
+  ),
+  "temp"
+);
+
+process.env["RUNNER_TOOL_CACHE"] = toolDir;
+process.env["RUNNER_TEMP"] = tempDir;
```
Comment (axetroy) [inline __tests__/installer.test.ts]: Same reason as above?

File: __tests__/installer.test.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      // This will throw if it doesn't find it in the cache (because no such version exists)
+      await installer.getDeno("251.0.0");
+    } catch {
+      thrown = true;
+    }
+    expect(thrown).toBe(true);
+    return;
+  });
+
+  it("Resolves semantic versions of node installed in cache", async () => {
```
Comment (axetroy) [inline __tests__/installer.test.ts]: > versions of node
versions of **deno**

File: package.json [hunk 5]
```diff
@@ -1,43 +1,53 @@
...
+  "name": "setup-deno",
+  "version": "1.1.0",
+  "private": true,
+  "description": "setup deno action",
+  "main": "src/setup-deno.ts",
+  "scripts": {
+    "build": "tsc",
+    "format": "prettier --write **/*.ts",
+    "format-check": "prettier --check **/*.ts",
+    "release": "ncc build && git add -f dist/",
```
Comment (axetroy) [inline package.json]: Nice 👍

File: .github/workflows/workflow.yml [hunk 6]
```diff
@@ -1,33 +1,33 @@
...
-        uses: ./
-        with:
-          deno-version: v0.18.0
-
-      - name: Run deno
-        run: |
-          which deno
-          deno https://deno.land/welcome.ts
+name: Main workflow
+on: [push]
```
Comment (axetroy) [inline .github/workflows/workflow.yml]: I think it's best to rename this file.

`workflow.yml` -> `ci.yml`
`name: Main workflow` -> `name: ci`

add add `pull_request` event

`on: [push]` -> `on: [push, pull_request]`

and don't forget update badge in `README.md`

File: src/release.json (hunk 2 above)
Comment (zhmushan) [reply src/release.json]: https://github.com/zhmushan/setup-deno/blob/refactor/src/installer.ts#L125
I think we should fetch `release.json` from web finally. If not, once deno updated, we need to update too

Comment (axetroy) [reply src/release.json]: Yeah.

that is next step. We should put this file into `/docs` folder and enable `gh-pages`

When this file is finalized, another PR can be opened

add files:

-/docs/release.json
-/docs/README.md

then load release.json file with http in this PR

File: __tests__/installer.test.ts (hunk 4 above)
Comment (zhmushan) [reply __tests__/installer.test.ts]: done.

File: .github/workflows/workflow.yml (hunk 6 above)
Comment (zhmushan) [reply .github/workflows/workflow.yml]: done.

File: src/release.json (hunk 2 above)
Comment (axetroy) [reply src/release.json]: could you open another PR for this?

after the `release.json` file is merged into master

then consider this PR again

Comment (zhmushan) [reply src/release.json]: Maybe we can use `now.sh` to deploy `git page` to forward data from `GitHub API`. This way we can avoid having to update release.json every time

Comment (axetroy) [reply src/release.json]: Even if you deploy with now, you still need to update release.json

I don’t think it is necessary to rely on third-party services

`gh-pages` is enough

Comment (zhmushan) [reply src/release.json]: #5

Comment (zhmushan) [reply src/release.json]: done.

**Comment by axetroy** [comment]
last step. create a test repository make sure it work.

then merge

**Comment by axetroy** [comment]
Am i missing something?

ci throw an 404 error when I try

https://github.com/axetroy/test_setup_deno/runs/322682828

**Comment by zhmushan** [comment]
> Am i missing something?
> 
> ci throw an 404 error when I try
> 
> https://github.com/axetroy/test_setup_deno/runs/322682828
> 
> maybe we should improve error message

https://github.com/axetroy/test_setup_deno/blob/master/.github/workflows/ci.yml#L15
`zhmushan/setup-deno@refactor`

File: src/installer.ts [hunk 7]
```diff
@@ -1,137 +1,159 @@
...
+  // Download - a tool installer intimately knows how to get the tool (and construct urls)
+  //
+  const c = semver.clean(version);
+  if (c) {
+    version = c;
+  } else {
+    throw new Error(`Unable to find Deno version ${version}`);
+  }
+  const fileName = `deno_${osPlat()}_${osArch()}`;
+  const urlFileName = osPlat() == "win" ? `${fileName}.zip` : `${fileName}.gz`;
```
Comment (axetroy) [inline src/installer.ts]: missing `x_64` suffix

it should be `deno_osx_x64.gz`

**Pull request merged**